src/ray/raylet/scheduling/cluster_resource_data.h

```diff
@@ -27,139 +27,546 @@
+  TaskResourceInstances(const ResourceRequest &request) {
+    for (size_t i = 0; i < PredefinedResourcesEnum_MAX; i++) {
+      this->predefined_resources_.push_back({});
     }
-  };
-  /// Get GPU instances only.
-  std::vector<FixedPoint> GetGPUInstances() const {
-    if (!this->predefined_resources.empty()) {
-      return this->predefined_resources[GPU];
+    for (auto &resource_id : request.ResourceIds()) {
+      std::vector<FixedPoint> instances;
+      auto value = request.Get(resource_id);
+      if (resource_id.IsUnitInstanceResource()) {
+        size_t num_instances = static_cast<size_t>(value.Double());
+        for (size_t i = 0; i < num_instances; i++) {
+          instances.push_back(1.0);
+        };
+      } else {
+        instances.push_back(value);
+      }
+      Set(resource_id, instances);
+    }
+  }
+
+  /// Get the per-instance values of a particular resource.
+  /// NOTE: the resource MUST already exist in this TaskResourceInstances, otherwise a
+  /// check fail will occur.
+  const std::vector<FixedPoint> &Get(const ResourceID resource_id) const {
+    auto ptr = GetPointer(resource_id);
+    RAY_CHECK(ptr != nullptr) << "Resource ID not found " << resource_id;
+    return *ptr;
+  }
+
+  /// Get the per-instance double values of a particular resource.
+  /// NOTE: the resource MUST already exist in this TaskResourceInstances, otherwise a
+  /// check fail will occur.
+  std::vector<double> GetDouble(const ResourceID resource_id) const {
+    return FixedPointVectorToDouble(Get(resource_id));
+  }
+
+  /// Get the sum of per-instance values of a particular resource.
+  /// If the resource doesn't exist, return 0.
+  FixedPoint Sum(const ResourceID resource_id) const {
+    if (Has(resource_id)) {
+      return FixedPoint::Sum(Get(resource_id));
     } else {
-      return {};
+      return FixedPoint(0);
     }
-  };
-  std::vector<double> GetGPUInstancesDouble() const {
-    if (!this->predefined_resources.empty()) {
-      return VectorFixedPointToVectorDouble(this->predefined_resources[GPU]);
+  }
+
+  /// Get the mutable per-instance values of a particular resource.
+  /// NOTE: the resource MUST already exist in this TaskResourceInstances, otherwise a
+  /// check fail will occur.
+  /// TODO(hchen): We should hide this method, and encapsulate all mutation operations.
+  std::vector<FixedPoint> &GetMutable(const ResourceID resource_id) {
+    auto ptr = GetPointer(resource_id);
+    RAY_CHECK(ptr != nullptr) << "Resource ID not found " << resource_id;
+    return *ptr;
+  }
+
+  /// Check whether a particular resource exists.
+  bool Has(ResourceID resource_id) const {
+    auto ptr = GetPointer(resource_id);
+    return ptr != nullptr && ptr->size() > 0;
+  }
+
+  /// Set the per-instance values for a particular resource.
+  TaskResourceInstances &Set(const ResourceID resource_id,
+                             const std::vector<FixedPoint> &instances) {
+    if (instances.size() == 0) {
```

Comment:
Can this ever happen?

Reply:
I'm not sure if we already have such a use case. But from this API's perspective, it's allowed. It's equivalent to "ResourceRequest::Set(resource_id, 0)".